mempool/reactor.go

```diff
@@ -136,9 +152,20 @@ func (memR *MempoolReactor) broadcastTxRoutine(peer p2p.Peer) {
 			time.Sleep(peerCatchupSleepIntervalMS * time.Millisecond)
 			continue
 		}
+		next = waitWithCancel(next.NextWait, ctx)
+		if ctx.Err() != nil {
+			return
+		}
+	}
+}
 
-		next = next.NextWait()
-		continue
+func waitWithCancel(f func() *clist.CElement, ctx context.Context) *clist.CElement {
+	el := make(chan *clist.CElement, 1)
+	select {
+	case el <- f():
```

Comment:
This will just run f() blockingly, so this doesn't change much in terms of behavior.

Reply:
behaviour of clist, yes. but select will exit once ctx is cancelled

Reply:
I wrote tests. So I am pretty sure it works

Reply:
I'm 100% sure that it doesn't. :)

https://play.golang.org/p/gN21yO9IRs3

Reply:
sh*t. hmm..